pandas/_libs/hashtable_class_helper.pxi.in

```diff
@@ -1099,32 +1133,46 @@ cdef class StringHashTable(HashTable):
             const char *v
             const char **vecs
             khiter_t k
+            flatiter it = PyArray_IterNew(values)
 
         # these by-definition *must* be strings
         vecs = <const char **>malloc(n * sizeof(char *))
         if vecs is NULL:
             raise MemoryError()
         for i in range(n):
-            val = values[i]
+            val = PyArray_GETITEM(values, PyArray_ITER_DATA(it))
 
             if isinstance(val, str):
                 # GH#31499 if we have a np.str_ PyUnicode_AsUTF8 won't recognize
                 #  it as a str, even though isinstance does.
                 v = PyUnicode_AsUTF8(<str>val)
             else:
                 v = PyUnicode_AsUTF8(self.na_string_sentinel)
+
+            # Need to copy result from PyUnicode_AsUTF8 when we have
+            # numpy strings
+            # Since numpy strings aren't backed by object arrays
+            # the buffer returned by PyUnicode_AsUTF8 will get freed
+            # in the next iteration when the created str object is GC'ed,
+            # clobbering the value of v
+            #if values.dtype.kind == "T":
+            v = strdup(v)
```

Comment:
This probably leaks in the current implementation

Reply:
I should be freeing these strings in ``__dealloc__`` if I didn't mess this up.

Reply:
No I wouldn't put it there either - `__dealloc__` is the inverse of `__cinit__`; any memory allocations performed outside of those functions needs to be managed with its own explicit lifecycle